### PR TITLE
GH-259 Setting default value for external_dependencies_patterns field…

### PR DESCRIPTION
**Pull request by maheshjfrog**
… in helm docker remote repo

Fixes: #267

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,10 @@
+## 2.23.2 (Mar 16, 2022)
+
+IMPROVEMENTS:
+
+* resource/artifactory_remote_docker_repository: Setting default value '**' for external_dependencies_patterns field. [GH-363]
```
Comment (chb0github) [inline CHANGELOG.md:5]: Hmmm, there is a local v1 and local v2 docker repo. Are you sure you don't need something identical for remote?

File: docs/resources/artifactory_remote_helm_repository.md [hunk 2]
```diff
@@ -43,9 +43,9 @@ All generic repo arguments are supported, in addition to:
 * `helm_charts_base_url` - (Optional) - No documentation is available. Hopefully you know what this means
 * `external_dependencies_enabled` - (Optional) When set, external dependencies are rewritten.
 * `external_dependencies_patterns` - (Optional) An Allow List of Ant-style path expressions that specify where external
-  dependencies may be downloaded from. By default, this is an empty list which means that no dependencies may be downloaded
-  from external sources. Note that the official documentation states the default is '**', which is correct when creating
-  repositories in the UI, but incorrect for the API.
+  dependencies may be downloaded from. By default, this is set to ** which means that dependencies may be downloaded
+  from any external source. Note that the official documentation states the default is '**', which is correct when creating
+  repositories in the UI, but incorrect for the API. This provider behavior will match to UI.
```
Comment (chb0github) [inline docs/resources/artifactory_remote_helm_repository.md:48]: Ditch the caveat - how we implement out backend is irrelevant to the user

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go [hunk 3]
```diff
@@ -32,9 +32,10 @@ func resourceArtifactoryRemoteHelmRepository() *schema.Resource {
 			},
 			RequiredWith: []string{"external_dependencies_enabled"},
 			Description: "An Allow List of Ant-style path expressions that specify where external dependencies may be downloaded from. " +
-				"By default, this is an empty list which means that no dependencies may be downloaded from external sources. " +
+				"By default, this is set to ** which means that dependencies may be downloaded from any external source. " +
 				"Note that the official documentation states the default is '**', " +
-				"which is correct when creating repositories in the UI, but incorrect for the API.",
+				"which is correct when creating repositories in the UI, but incorrect for the API. " +
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_remote_helm_repository.go:37]: This last bit is our implementation detail. Delete it

File: CHANGELOG.md (hunk 1 above)
Comment (maheshjfrog) [reply CHANGELOG.md:5]: For docker remote repos ... only v2 is present.

File: docs/resources/artifactory_remote_helm_repository.md [hunk 4]
```diff
@@ -43,8 +43,8 @@ All generic repo arguments are supported, in addition to:
 * `helm_charts_base_url` - (Optional) - No documentation is available. Hopefully you know what this means
 * `external_dependencies_enabled` - (Optional) When set, external dependencies are rewritten.
 * `external_dependencies_patterns` - (Optional) An Allow List of Ant-style path expressions that specify where external
-  dependencies may be downloaded from. By default, this is an empty list which means that no dependencies may be downloaded
-  from external sources. Note that the official documentation states the default is '**', which is correct when creating
+  dependencies may be downloaded from. By default, this is set to ** which means that dependencies may be downloaded
+  from any external source. Note that the official documentation states the default is '**', which is correct when creating
```
Comment (chb0github) [inline docs/resources/artifactory_remote_helm_repository.md:47]: Literally delete line 47
```suggestion

```

File: pkg/artifactory/resource_artifactory_remote_docker_repository.go [hunk 5]
```diff
@@ -56,10 +56,21 @@ func resourceArtifactoryRemoteDockerRepository() *schema.Resource {
 			BlockPushingSchema1:          d.getBool("block_pushing_schema1", false),
 			ExternalDependenciesPatterns: d.getList("external_dependencies_patterns"),
 		}
+		if len(repo.ExternalDependenciesPatterns) == 0 {
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_remote_docker_repository.go:59]: Actually, take a look at what @alexhung did in `resource_artifactory_ldap_settings`

```go
		"email_attribute": {
			Type:     schema.TypeString,
			Optional: true,
			Default:  "mail",
			StateFunc: func(str interface{}) string {
				// To match Artifactory behavior, Setting "email_attribute" value to "mail" when blank input is specified.
				value, ok := str.(string)
				if !ok || value == "" {
					return "mail"
				}
				return value
			},
			Description: `(Optional) An attribute that can be used to map a user's email address to a user created automatically in Artifactory. Default value is "mail".`,
		},
```

This is a more idiomatic approach

File: pkg/artifactory/resource_artifactory_remote_docker_repository.go [hunk 6]
```diff
@@ -56,10 +56,21 @@ func resourceArtifactoryRemoteDockerRepository() *schema.Resource {
...
+			repo.ExternalDependenciesPatterns = []string{"**"}
+		}
 		return repo, repo.Id(), nil
 	}
 
-	return mkResourceSchema(dockerRemoteSchema, defaultPacker, unpackDockerRemoteRepo, func() interface{} {
+	// Special handling for "external_dependencies_patterns" attribute to match default value behavior in UI.
+	dockerRemoteRepoPacker := universalPack(
+		allHclPredicate(
+			ignoreHclPredicate("class", "rclass", "external_dependencies_patterns"),
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_remote_docker_repository.go:68]: Yeah, not a fan of this - see if you can remove this if you use the `StateFunc`

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go [hunk 7]
```diff
@@ -32,7 +32,7 @@ func resourceArtifactoryRemoteHelmRepository() *schema.Resource {
 			},
 			RequiredWith: []string{"external_dependencies_enabled"},
 			Description: "An Allow List of Ant-style path expressions that specify where external dependencies may be downloaded from. " +
-				"By default, this is an empty list which means that no dependencies may be downloaded from external sources. " +
+				"By default, this is set to ** which means that dependencies may be downloaded from any external source. " +
 				"Note that the official documentation states the default is '**', " +
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_remote_helm_repository.go:36]:
```suggestion

```

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go [hunk 8]
```diff
@@ -32,7 +32,7 @@ func resourceArtifactoryRemoteHelmRepository() *schema.Resource {
...
 				"which is correct when creating repositories in the UI, but incorrect for the API.",
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_remote_helm_repository.go:37]:
```suggestion

```

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go [hunk 9]
```diff
@@ -53,10 +53,21 @@ func resourceArtifactoryRemoteHelmRepository() *schema.Resource {
 			ExternalDependenciesEnabled:  d.getBool("external_dependencies_enabled", false),
 			ExternalDependenciesPatterns: d.getList("external_dependencies_patterns"),
 		}
+		if len(repo.ExternalDependenciesPatterns) == 0 {
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_remote_helm_repository.go:56]: try managing this in the state func

**Comment by maheshjfrog** [comment]
Unable to use StateFunc for this scenario.

**Review by chb0github — APPROVED** [review]
If this isn't possible in the state func, then so be it

**Pull request merged**